<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.14: http://docutils.sourceforge.net/" />
<title>Compound-Protein Interaction Pipeline</title>
<style type="text/css">

/*
:Author: David Goodger (goodger@python.org)
:Id: $Id: html4css1.css 7952 2016-07-26 18:15:59Z milde $
:Copyright: This stylesheet has been placed in the public domain.

Default cascading style sheet for the HTML output of Docutils.

See http://docutils.sf.net/docs/howto/html-stylesheets.html for how to
customize this style sheet.
*/

/* used to remove borders from tables and images */
.borderless, table.borderless td, table.borderless th {
  border: 0 }

table.borderless td, table.borderless th {
  /* Override padding for "table.docutils td" with "! important".
     The right padding separates the table cells. */
  padding: 0 0.5em 0 0 ! important }

.first {
  /* Override more specific margin styles with "! important". */
  margin-top: 0 ! important }

.last, .with-subtitle {
  margin-bottom: 0 ! important }

.hidden {
  display: none }

.subscript {
  vertical-align: sub;
  font-size: smaller }

.superscript {
  vertical-align: super;
  font-size: smaller }

a.toc-backref {
  text-decoration: none ;
  color: black }

blockquote.epigraph {
  margin: 2em 5em ; }

dl.docutils dd {
  margin-bottom: 0.5em }

object[type="image/svg+xml"], object[type="application/x-shockwave-flash"] {
  overflow: hidden;
}

/* Uncomment (and remove this text!) to get bold-faced definition list terms
dl.docutils dt {
  font-weight: bold }
*/

div.abstract {
  margin: 2em 5em }

div.abstract p.topic-title {
  font-weight: bold ;
  text-align: center }

div.admonition, div.attention, div.caution, div.danger, div.error,
div.hint, div.important, div.note, div.tip, div.warning {
  margin: 2em ;
  border: medium outset ;
  padding: 1em }

div.admonition p.admonition-title, div.hint p.admonition-title,
div.important p.admonition-title, div.note p.admonition-title,
div.tip p.admonition-title {
  font-weight: bold ;
  font-family: sans-serif }

div.attention p.admonition-title, div.caution p.admonition-title,
div.danger p.admonition-title, div.error p.admonition-title,
div.warning p.admonition-title, .code .error {
  color: red ;
  font-weight: bold ;
  font-family: sans-serif }

/* Uncomment (and remove this text!) to get reduced vertical space in
   compound paragraphs.
div.compound .compound-first, div.compound .compound-middle {
  margin-bottom: 0.5em }

div.compound .compound-last, div.compound .compound-middle {
  margin-top: 0.5em }
*/

div.dedication {
  margin: 2em 5em ;
  text-align: center ;
  font-style: italic }

div.dedication p.topic-title {
  font-weight: bold ;
  font-style: normal }

div.figure {
  margin-left: 2em ;
  margin-right: 2em }

div.footer, div.header {
  clear: both;
  font-size: smaller }

div.line-block {
  display: block ;
  margin-top: 1em ;
  margin-bottom: 1em }

div.line-block div.line-block {
  margin-top: 0 ;
  margin-bottom: 0 ;
  margin-left: 1.5em }

div.sidebar {
  margin: 0 0 0.5em 1em ;
  border: medium outset ;
  padding: 1em ;
  background-color: #ffffee ;
  width: 40% ;
  float: right ;
  clear: right }

div.sidebar p.rubric {
  font-family: sans-serif ;
  font-size: medium }

div.system-messages {
  margin: 5em }

div.system-messages h1 {
  color: red }

div.system-message {
  border: medium outset ;
  padding: 1em }

div.system-message p.system-message-title {
  color: red ;
  font-weight: bold }

div.topic {
  margin: 2em }

h1.section-subtitle, h2.section-subtitle, h3.section-subtitle,
h4.section-subtitle, h5.section-subtitle, h6.section-subtitle {
  margin-top: 0.4em }

h1.title {
  text-align: center }

h2.subtitle {
  text-align: center }

hr.docutils {
  width: 75% }

img.align-left, .figure.align-left, object.align-left, table.align-left {
  clear: left ;
  float: left ;
  margin-right: 1em }

img.align-right, .figure.align-right, object.align-right, table.align-right {
  clear: right ;
  float: right ;
  margin-left: 1em }

img.align-center, .figure.align-center, object.align-center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}

table.align-center {
  margin-left: auto;
  margin-right: auto;
}

.align-left {
  text-align: left }

.align-center {
  clear: both ;
  text-align: center }

.align-right {
  text-align: right }

/* reset inner alignment in figures */
div.align-right {
  text-align: inherit }

/* div.align-center * { */
/*   text-align: left } */

.align-top    {
  vertical-align: top }

.align-middle {
  vertical-align: middle }

.align-bottom {
  vertical-align: bottom }

ol.simple, ul.simple {
  margin-bottom: 1em }

ol.arabic {
  list-style: decimal }

ol.loweralpha {
  list-style: lower-alpha }

ol.upperalpha {
  list-style: upper-alpha }

ol.lowerroman {
  list-style: lower-roman }

ol.upperroman {
  list-style: upper-roman }

p.attribution {
  text-align: right ;
  margin-left: 50% }

p.caption {
  font-style: italic }

p.credits {
  font-style: italic ;
  font-size: smaller }

p.label {
  white-space: nowrap }

p.rubric {
  font-weight: bold ;
  font-size: larger ;
  color: maroon ;
  text-align: center }

p.sidebar-title {
  font-family: sans-serif ;
  font-weight: bold ;
  font-size: larger }

p.sidebar-subtitle {
  font-family: sans-serif ;
  font-weight: bold }

p.topic-title {
  font-weight: bold }

pre.address {
  margin-bottom: 0 ;
  margin-top: 0 ;
  font: inherit }

pre.literal-block, pre.doctest-block, pre.math, pre.code {
  margin-left: 2em ;
  margin-right: 2em }

pre.code .ln { color: grey; } /* line numbers */
pre.code, code { background-color: #eeeeee }
pre.code .comment, code .comment { color: #5C6576 }
pre.code .keyword, code .keyword { color: #3B0D06; font-weight: bold }
pre.code .literal.string, code .literal.string { color: #0C5404 }
pre.code .name.builtin, code .name.builtin { color: #352B84 }
pre.code .deleted, code .deleted { background-color: #DEB0A1}
pre.code .inserted, code .inserted { background-color: #A3D289}

span.classifier {
  font-family: sans-serif ;
  font-style: oblique }

span.classifier-delimiter {
  font-family: sans-serif ;
  font-weight: bold }

span.interpreted {
  font-family: sans-serif }

span.option {
  white-space: nowrap }

span.pre {
  white-space: pre }

span.problematic {
  color: red }

span.section-subtitle {
  /* font-size relative to parent (h1..h6 element) */
  font-size: 80% }

table.citation {
  border-left: solid 1px gray;
  margin-left: 1px }

table.docinfo {
  margin: 2em 4em }

table.docutils {
  margin-top: 0.5em ;
  margin-bottom: 0.5em }

table.footnote {
  border-left: solid 1px black;
  margin-left: 1px }

table.docutils td, table.docutils th,
table.docinfo td, table.docinfo th {
  padding-left: 0.5em ;
  padding-right: 0.5em ;
  vertical-align: top }

table.docutils th.field-name, table.docinfo th.docinfo-name {
  font-weight: bold ;
  text-align: left ;
  white-space: nowrap ;
  padding-left: 0 }

/* "booktabs" style (no vertical lines) */
table.docutils.booktabs {
  border: 0px;
  border-top: 2px solid;
  border-bottom: 2px solid;
  border-collapse: collapse;
}
table.docutils.booktabs * {
  border: 0px;
}
table.docutils.booktabs th {
  border-bottom: thin solid;
  text-align: left;
}

h1 tt.docutils, h2 tt.docutils, h3 tt.docutils,
h4 tt.docutils, h5 tt.docutils, h6 tt.docutils {
  font-size: 100% }

ul.auto-toc {
  list-style-type: none }

</style>
</head>
<body>
<div class="document" id="compound-protein-interaction-pipeline">
<h1 class="title">Compound-Protein Interaction Pipeline</h1>

<div class="section" id="introduction">
<h1>Introduction</h1>
<ul>
<li><p class="first">The presented pipeline uses the protein-protein interaction benchmark package published by Tikk et al. (Tikk et al., 2010. A comprehensive benchmark of kernel methods to extract protein-protein interactions from literature. PLoS Comput. Biol).</p>
</li>
<li><p class="first">The selection of considered kernels and software components was reduced to the all-paths graph (APG) kernel and the shallow linguistic (SL) kernel, because they performed best on the protein-protein interaction (PPI) data sets.</p>
</li>
<li><p class="first">In this project, the two kernels were applied to two compound-protein interaction (CPI) in PubMed sentences.</p>
</li>
<li><p class="first">The complete data set consists of 2753 sentences. Within these sentences, 3496 compound-protein pairs were annotated to show a functional relationship and 2467 compound-protein pairs were marked as no-interaction pairs.</p>
</li>
<li><p class="first">We consider co-occurences as an approach to predict every appearance of a compound and a protein in a sentence as a functional relationship (recall 100%, specificity 0%) and call it the baseline calculation, taking into account the number of true functional relationships.</p>
</li>
<li><p class="first">From the given numbers, the data set shows an precision (equal to accuracy in this case) of 58.6 % and an F1 score of 73.9 %.</p>
</li>
<li><p class="first">APG and SL kernel both perform better than the concept of co-occurrences as shown in the following tables.</p>
</li>
<li><p class="first">APG kernel results for CPI-DS dataset:</p>
<blockquote>
<table border="1" class="colwidths-auto docutils">
<thead valign="bottom">
<tr><th class="head">Param.</th>
<th class="head">Sens.</th>
<th class="head">Spec.</th>
<th class="head">Prec.</th>
<th class="head">Acc.</th>
<th class="head">F1</th>
<th class="head">AUC</th>
</tr>
</thead>
<tbody valign="top">
<tr><td>0.25</td>
<td>85.0</td>
<td>65.3</td>
<td>77.8</td>
<td>77.0</td>
<td>81.2</td>
<td>84.1</td>
</tr>
<tr><td>0.50</td>
<td>84.6</td>
<td>65.8</td>
<td>77.9</td>
<td>77.0</td>
<td>81.1</td>
<td>84.1</td>
</tr>
<tr><td>1.00</td>
<td>84.0</td>
<td>67.0</td>
<td>78.4</td>
<td>77.2</td>
<td>81.1</td>
<td>83.8</td>
</tr>
<tr><td>2.00</td>
<td>84.2</td>
<td>66.4</td>
<td>78.1</td>
<td>77.0</td>
<td>81.0</td>
<td>83.4</td>
</tr>
</tbody>
</table>
</blockquote>
</li>
<li><p class="first">SL kernel results for CPI-DS dataset:</p>
<blockquote>
<table border="1" class="colwidths-auto docutils">
<thead valign="bottom">
<tr><th class="head">n</th>
<th class="head">w</th>
<th class="head">Sens.</th>
<th class="head">Spec.</th>
<th class="head">Prec.</th>
<th class="head">Acc.</th>
<th class="head">F1</th>
<th class="head">AUC</th>
</tr>
</thead>
<tbody valign="top">
<tr><td>1</td>
<td>1</td>
<td>77.1</td>
<td>68.6</td>
<td>77.9</td>
<td>73.6</td>
<td>77.4</td>
<td>79.1</td>
</tr>
<tr><td>1</td>
<td>2</td>
<td>83.0</td>
<td>61.1</td>
<td>75.4</td>
<td>74.0</td>
<td>78.9</td>
<td>78.6</td>
</tr>
<tr><td>1</td>
<td>3</td>
<td>87.1</td>
<td>53.0</td>
<td>72.5</td>
<td>73.0</td>
<td>79.1</td>
<td>78.1</td>
</tr>
<tr><td>2</td>
<td>1</td>
<td>79.0</td>
<td>68.8</td>
<td>78.3</td>
<td>74.8</td>
<td>78.6</td>
<td>80.9</td>
</tr>
<tr><td>2</td>
<td>2</td>
<td>84.5</td>
<td>60.8</td>
<td>75.5</td>
<td>74.7</td>
<td>79.7</td>
<td>80.3</td>
</tr>
<tr><td>2</td>
<td>3</td>
<td>87.9</td>
<td>54.8</td>
<td>73.7</td>
<td>74.3</td>
<td>80.1</td>
<td>79.9</td>
</tr>
<tr><td>3</td>
<td>1</td>
<td>80.3</td>
<td>68.7</td>
<td>78.5</td>
<td>75.5</td>
<td>79.3</td>
<td>81.4</td>
</tr>
<tr><td>3</td>
<td>2</td>
<td>84.4</td>
<td>62.0</td>
<td>76.1</td>
<td>75.2</td>
<td>80.0</td>
<td>80.8</td>
</tr>
<tr><td>3</td>
<td>3</td>
<td>88.2</td>
<td>56.3</td>
<td>74.3</td>
<td>75.1</td>
<td>80.6</td>
<td>80.6</td>
</tr>
</tbody>
</table>
</blockquote>
</li>
</ul>
<div class="section" id="functional-relationships-with-and-without-an-enclosed-interaction-verb">
<h2>Functional relationships with and without an enclosed interaction verb</h2>
<ul>
<li><p class="first">The whole prediction can be further divided into considering only pairs of compounds and proteins which enclose an interaction verb (CPI-DS_IV) and pairs of compounds and proteins that don't show this sentence structure (CPI-DS_NIV).</p>
</li>
<li><p class="first">These verbs have been defined in the publication of the web service prolific (Senger and Grüning et al., 2012. Mining and evaluation of molecular relationships in literature. Bioinformatics).</p>
</li>
<li><p class="first">The following table shows the evaluation of the co-occurrences approach (DS - Data set, Sent. - Sentences, Sens. - Sensitivity, Spec. - Specificity, Prec. - Precision, Acc. - Accuracy, F1 - F1 score, evaluation parameters shown in percent):</p>
<blockquote>
<table border="1" class="colwidths-auto docutils">
<thead valign="bottom">
<tr><th class="head">DS</th>
<th class="head">#Sent.</th>
<th class="head">#CPIs</th>
<th class="head">#No-CPIs</th>
<th class="head">Total</th>
<th class="head">Sens.</th>
<th class="head">Spec.</th>
<th class="head">Prec.</th>
<th class="head">F1</th>
</tr>
</thead>
<tbody valign="top">
<tr><td>CPI-DS_IV</td>
<td>1259</td>
<td>1884</td>
<td>1166</td>
<td>3050</td>
<td>100.0</td>
<td>0.0</td>
<td>61.8</td>
<td>76.4</td>
</tr>
<tr><td>CPI-DS_NIV</td>
<td>1494</td>
<td>1612</td>
<td>1301</td>
<td>2913</td>
<td>100.0</td>
<td>0.0</td>
<td>55.3</td>
<td>71.2</td>
</tr>
</tbody>
</table>
</blockquote>
</li>
<li><p class="first">The next table shows the results of the APG kernel pipeline for CPI-DS_IV dataset:</p>
<blockquote>
<table border="1" class="colwidths-auto docutils">
<thead valign="bottom">
<tr><th class="head">Param.</th>
<th class="head">Sens.</th>
<th class="head">Spec.</th>
<th class="head">Prec.</th>
<th class="head">Acc.</th>
<th class="head">F1</th>
<th class="head">AUC</th>
</tr>
</thead>
<tbody valign="top">
<tr><td>0.25</td>
<td>85.4</td>
<td>59.6</td>
<td>78.4</td>
<td>76.3</td>
<td>81.7</td>
<td>81.7</td>
</tr>
<tr><td>0.50</td>
<td>85.0</td>
<td>59.8</td>
<td>78.4</td>
<td>76.1</td>
<td>81.5</td>
<td>81.6</td>
</tr>
<tr><td>1.00</td>
<td>84.4</td>
<td>60.9</td>
<td>79.0</td>
<td>76.3</td>
<td>81.6</td>
<td>81.4</td>
</tr>
<tr><td>2.00</td>
<td>84.6</td>
<td>60.1</td>
<td>78.5</td>
<td>76.0</td>
<td>81.4</td>
<td>81.0</td>
</tr>
</tbody>
</table>
</blockquote>
</li>
<li><p class="first">Results of the APG kernel pipeline for CPI-DS_NIV dataset:</p>
<blockquote>
<table border="1" class="colwidths-auto docutils">
<thead valign="bottom">
<tr><th class="head">Param.</th>
<th class="head">Sens.</th>
<th class="head">Spec.</th>
<th class="head">Prec.</th>
<th class="head">Acc.</th>
<th class="head">F1</th>
<th class="head">AUC</th>
</tr>
</thead>
<tbody valign="top">
<tr><td>0.25</td>
<td>84.3</td>
<td>69.0</td>
<td>77.3</td>
<td>77.5</td>
<td>80.5</td>
<td>85.6</td>
</tr>
<tr><td>0.50</td>
<td>84.1</td>
<td>69.9</td>
<td>77.7</td>
<td>77.8</td>
<td>80.6</td>
<td>85.5</td>
</tr>
<tr><td>1.00</td>
<td>83.5</td>
<td>70.9</td>
<td>78.1</td>
<td>77.8</td>
<td>80.6</td>
<td>85.4</td>
</tr>
<tr><td>2.00</td>
<td>83.7</td>
<td>70.2</td>
<td>77.8</td>
<td>77.7</td>
<td>80.6</td>
<td>85.0</td>
</tr>
</tbody>
</table>
</blockquote>
</li>
<li><p class="first">This table shows the results of the SL kernel for CPI-DS_IV dataset:</p>
<blockquote>
<table border="1" class="colwidths-auto docutils">
<thead valign="bottom">
<tr><th class="head">n</th>
<th class="head">w</th>
<th class="head">Sens.</th>
<th class="head">Spec.</th>
<th class="head">Prec.</th>
<th class="head">Acc.</th>
<th class="head">F1</th>
<th class="head">AUC</th>
</tr>
</thead>
<tbody valign="top">
<tr><td>1</td>
<td>1</td>
<td>78.8</td>
<td>60.4</td>
<td>76.5</td>
<td>71.9</td>
<td>77.5</td>
<td>75.5</td>
</tr>
<tr><td>1</td>
<td>2</td>
<td>82.2</td>
<td>55.2</td>
<td>75.1</td>
<td>72.1</td>
<td>78.3</td>
<td>74.8</td>
</tr>
<tr><td>1</td>
<td>3</td>
<td>86.7</td>
<td>46.7</td>
<td>72.6</td>
<td>71.6</td>
<td>78.9</td>
<td>74.6</td>
</tr>
<tr><td>2</td>
<td>1</td>
<td>79.9</td>
<td>61.1</td>
<td>77.1</td>
<td>72.9</td>
<td>78.3</td>
<td>77.3</td>
</tr>
<tr><td>2</td>
<td>2</td>
<td>84.0</td>
<td>54.2</td>
<td>74.9</td>
<td>72.7</td>
<td>79.0</td>
<td>76.7</td>
</tr>
<tr><td>2</td>
<td>3</td>
<td>88.3</td>
<td>47.6</td>
<td>73.2</td>
<td>72.8</td>
<td>79.9</td>
<td>76.4</td>
</tr>
<tr><td>3</td>
<td>1</td>
<td>81.1</td>
<td>61.7</td>
<td>77.8</td>
<td>74.0</td>
<td>79.3</td>
<td>78.3</td>
</tr>
<tr><td>3</td>
<td>2</td>
<td>83.5</td>
<td>56.8</td>
<td>76.2</td>
<td>73.7</td>
<td>79.6</td>
<td>77.5</td>
</tr>
<tr><td>3</td>
<td>3</td>
<td>88.5</td>
<td>50.7</td>
<td>74.7</td>
<td>74.3</td>
<td>80.9</td>
<td>77.4</td>
</tr>
</tbody>
</table>
</blockquote>
</li>
<li><p class="first">Results of the SL kernel for CPI-DS_NIV dataset:</p>
<blockquote>
<table border="1" class="colwidths-auto docutils">
<thead valign="bottom">
<tr><th class="head">n</th>
<th class="head">w</th>
<th class="head">Sens.</th>
<th class="head">Spec.</th>
<th class="head">Prec.</th>
<th class="head">Acc.</th>
<th class="head">F1</th>
<th class="head">AUC</th>
</tr>
</thead>
<tbody valign="top">
<tr><td>1</td>
<td>1</td>
<td>75.5</td>
<td>74.8</td>
<td>78.9</td>
<td>75.3</td>
<td>77.1</td>
<td>81.9</td>
</tr>
<tr><td>1</td>
<td>2</td>
<td>82.7</td>
<td>67.3</td>
<td>76.0</td>
<td>75.9</td>
<td>79.1</td>
<td>81.4</td>
</tr>
<tr><td>1</td>
<td>3</td>
<td>87.4</td>
<td>58.3</td>
<td>72.3</td>
<td>74.4</td>
<td>79.0</td>
<td>80.9</td>
</tr>
<tr><td>2</td>
<td>1</td>
<td>78.3</td>
<td>75.0</td>
<td>79.6</td>
<td>76.8</td>
<td>78.8</td>
<td>83.6</td>
</tr>
<tr><td>2</td>
<td>2</td>
<td>84.2</td>
<td>69.0</td>
<td>77.4</td>
<td>77.4</td>
<td>80.5</td>
<td>83.1</td>
</tr>
<tr><td>2</td>
<td>3</td>
<td>87.6</td>
<td>62.5</td>
<td>74.5</td>
<td>76.4</td>
<td>80.4</td>
<td>82.9</td>
</tr>
<tr><td>3</td>
<td>1</td>
<td>79.7</td>
<td>74.5</td>
<td>79.6</td>
<td>77.4</td>
<td>79.5</td>
<td>84.0</td>
</tr>
<tr><td>3</td>
<td>2</td>
<td>84.2</td>
<td>68.6</td>
<td>77.1</td>
<td>77.2</td>
<td>80.4</td>
<td>83.5</td>
</tr>
<tr><td>3</td>
<td>3</td>
<td>88.6</td>
<td>62.4</td>
<td>74.7</td>
<td>76.9</td>
<td>80.9</td>
<td>83.3</td>
</tr>
</tbody>
</table>
</blockquote>
</li>
<li><p class="first">The ratio of sentences with and without interaction verbs for the complete PubMed data set is around 40 % CPI-DS_IV and 60 % CPI-DS_NIV, based on an analysis with PubMedPortable.</p>
</li>
<li><p class="first">In the case of CPI-DS_IV and CPI-DS_NIV, the percentages are 45.7 % (1259/(1259+1494)) and 54.3 % (1494/(1259+1494)).</p>
</li>
<li><p class="first">Considering the manual curation of false positives resulting from the automatic named entity recognition process, the empirical ratios can be considered as similar.</p>
</li>
<li><p class="first">The predictions for CPI-DS_IV and CPI-DS_NIV are based on training the whole model of CPI-DS, using the same splits. Therefore, evaluation is separated by the two different sentence structures, but the overall model stays the same for CPI-DS, CPI-DS_IV, and CPI-DS_NIV, regarding each cross-validation run.</p>
</li>
</ul>
</div>
<div class="section" id="the-benchmark-data-set-creation">
<h2>The benchmark data set creation</h2>
<ul>
<li><p class="first">The data sets were created by selecting the first 40,000 PubMed abstracts from 2009 with PubMedPortable (<a class="reference external" href="https://github.com/KerstenDoering/PubMedPortable">https://github.com/KerstenDoering/PubMedPortable</a>).</p>
</li>
<li><p class="first">The chemical compounds were identified with the backend of prolific using the Hettne rules (Hettne et al., 2009. A dictionary to identify small molecules and drugs in free text. Bioinformatics).</p>
</li>
<li><p class="first">The web service Whatizit was used for gene and protein synonyms (Rebholz-Schuhmann,D. et al., 2008. Text processing through Web services: calling Whatizit. Bioinformatics).</p>
</li>
<li><p class="first">After automated extraction of all sentences with an interaction verb (first 20,000 abstracts for CPI-DS_IV) and without an interaction enclosed by two biomolecues (second 20,000 abstracts for CPI-DS_NIV), they were displayed in HTML pages with Javascript buttons to select a status for each sentence.</p>
</li>
<li><p class="first">If a sentence contained at least one interacting pair of biomolecules, the status &quot;Interaction&quot; was selected for this sentence. The alternative was &quot;No Interaction&quot;.</p>
</li>
<li><p class="first">If the named entity recognition tools tagged false positive synonyms, they were annotated in the separate file RM_comments.csv.</p>
</li>
<li><p class="first">If there were non-interacting molecules in a sentence with the status &quot;Interaction&quot;, they were also annotated in the separate CSV table.</p>
</li>
<li><p class="first">If every molecule in the sentence was a false positive named entity, the status of the sentence was set to &quot;False positive example&quot;.</p>
</li>
<li><p class="first">The data set curation was done by Michael Becer, supervised by Kersten Döring.</p>
</li>
<li><p class="first">All potentially functional interaction pairs marked as false positives were not considered in the evaluation process of the two kernel approaches.</p>
</li>
<li><p class="first">More information about the theoretical background of the kernels, the data sets, and other related information can be found in Kersten Döring's Dissertation (<a class="reference external" href="https://www.freidok.uni-freiburg.de/data/10565">https://www.freidok.uni-freiburg.de/data/10565</a>).</p>
<blockquote>
<ul class="simple">
<li>The SL kernel results contained in this thesis are not the results presented here, because they were generated with a customised pipeline, independently from the ppi-benchmark package.</li>
<li>The PubMedPortable project is referred to as PubMed2Go in this thesis.</li>
</ul>
</blockquote>
</li>
</ul>
</div>
<div class="section" id="technical-background-of-the-xml-data-set">
<h2>Technical background of the XML data set</h2>
<ul>
<li><p class="first">The script parser.py in the directories scripts/generate_XML_files/CPI-DS_IV/ und scripts/generate_XML_files/CPI-DS_NIV creates the files interactions.txt and no_interactions.txt from the HTML files in the folder data_sets.</p>
<blockquote>
<ul class="simple">
<li>The names of these files refer to the basic status of all sentences inside the text file.</li>
<li>Relationship_Mining_10000.html and Relationship_Mining_20000.html were concatenated to the file Relationship_Mining.html resulting in CPI-DS_IV. The other two HTML files represent CPI-DS_NIV.</li>
</ul>
</blockquote>
</li>
<li><p class="first">The script get_relations.py in the directories scripts/generate_XML_files/CPI-DS_IV/ und scripts/generate_XML_files/CPI-DS_NIV processes the files interaction.txt and no_interaction.txt using the (concatenated) file RM_comments.csv with annotations of false positive compounds (FP-C), false positive proteins (FP-P), non-interacting compounds (NI-C), and non-interacting proteins (NI-P).</p>
<blockquote>
<ul class="simple">
<li>The ouptut of get_relations.py is the file training_dataset.txt.</li>
<li>This file can be sorted by PubMed IDs with the command &quot;sort -u training_dataset.txt &gt; training_dataset_sorted.csv&quot;, individually done with CPI-DS_IV and CPI-DS_NIV.</li>
</ul>
</blockquote>
</li>
<li><p class="first">The script annotatedsen_to_xml.py generates the files CPI-DS_IV.xml and CPI-DS_NIV.xml, building a unified XML format described by Tikk et al.</p>
<blockquote>
<ul class="simple">
<li>This script was developed by Elham Abbasian in her Master Thesis, supervised by Kersten Döring.</li>
</ul>
</blockquote>
</li>
<li><p class="first">The SL kernel implementation in the ppi-benchmark package of Tikk et al. worked with this XML format, but the preprocessing for the APG kernel-required format did not work straight forward. Therefore, the single steps described in their documentation appendix (here: scripts/ppi-benchmark/documentationkernels-howto.pdf) were isolated from the available source code and stored as executable JAR files in the folder scripts/CPI-corpora-preparing with Eclipse.</p>
</li>
<li><p class="first">The XML files can be checked for consistency. Furthermore, it can be seen how many documents (PubMed IDs) and sentences there are:</p>
<blockquote>
<p>java -jar ConsistencyChecker.jar CPI-DS_IV/CPI-DS_IV.xml</p>
<blockquote>
<p>INFO: Processing 'CPI-DS_IV/CPI-DS_IV.xml'...</p>
<p>Processing corpus 'CPI-DS_IV' ...</p>
<p>Documents seen: 830</p>
<p>Sentences seen: 1259</p>
</blockquote>
<p>java -jar ConsistencyChecker.jar CPI-DS_NIV/CPI-DS_NIV.xml</p>
<blockquote>
<p>INFO: Processing 'CPI-DS_NIV/CPI-DS_NIV.xml'...</p>
<p>Processing corpus 'CPI-DS_NIV' ...</p>
<p>processed 1000 docs (2155.1724 docs/sec)</p>
<p>Documents seen: 1066</p>
<p>Sentences seen: 1494</p>
</blockquote>
<p>java -jar ConsistencyChecker.jar DS_40/DS-40.xml</p>
<blockquote>
<p>INFO: Processing 'DS_test_case_40_sentences/DS-40.xml'...</p>
<p>Processing corpus 'DS-40' ...</p>
<p>Documents seen: 26</p>
<p>Sentences seen: 40</p>
</blockquote>
</blockquote>
</li>
</ul>
</div>
</div>
<div class="section" id="technical-requirements-for-running-apg-and-sl-kernel-pipeline">
<h1>Technical Requirements for running APG and SL Kernel Pipeline</h1>
<ul class="simple">
<li>Many of the following steps are described similarly in the original ppi-benchmark documentation (scripts/ppi-benchmark/documentationkernels-howto.pdf).</li>
</ul>
<div class="section" id="required-installation-packages">
<h2>Required Installation Packages</h2>
<ul>
<li><p class="first">The operating system for the CPI-pipeline was Ubuntu 14.04 LTS and the whole pipeline was tested with Ubuntu 16.04 LTS &amp; Ubuntu 18.04 LTS.</p>
</li>
<li><p class="first">All packages can be installed with &quot;apt-get install&quot;, Synaptic Package Manager, or the new Ubuntu Software Center.</p>
</li>
<li><p class="first">List of packages:</p>
<blockquote>
<ul class="simple">
<li>Python 2.7</li>
<li>postgresql (here: version 9.3, used to store evaluation results)</li>
<li>libxml-perl (here: version 0.08, used for the generation of cross-validation splits)</li>
<li>python-lxml</li>
<li>python-numpy</li>
<li>default-jdk (Java 7 or 8, used for the execution of JAR files)</li>
<li>python-nltk</li>
<li>python-pip</li>
<li>pip install bllipparser (python bindings for the BLLIP natural language parser)</li>
</ul>
</blockquote>
</li>
</ul>
</div>
<div class="section" id="postgresql-configuration">
<h2>PostgreSQL Configuration</h2>
<ul>
<li><p class="first">Set your own user to be a PostgreSLQ superuser after installation of the default package &quot;postgresql&quot;.</p>
<blockquote>
<ul>
<li><p class="first">Follow the instructions here:</p>
<blockquote>
<p><a class="reference external" href="https://github.com/KerstenDoering/PubMedPortable/blob/master/documentation/quick_install.rst#creation-of-postgresql-superuser">https://github.com/KerstenDoering/PubMedPortable/blob/master/documentation/quick_install.rst#creation-of-postgresql-superuser</a></p>
</blockquote>
</li>
</ul>
</blockquote>
</li>
<li><p class="first">Type in the following commands to create a new user &quot;ppi&quot;.</p>
<blockquote>
<ul>
<li><p class="first">sudo useradd ppi -s /bin/false</p>
</li>
<li><p class="first">sudo passwd ppi</p>
</li>
<li><p class="first">sudo su -c &quot;psql&quot; postgres</p>
<blockquote>
<ul class="simple">
<li>CREATE USER ppi WITH PASSWORD 'ppi';</li>
<li>CREATE DATABASE ppi;</li>
<li>GRANT ALL PRIVILEGES ON DATABASE ppi TO ppi;</li>
<li>CREATE LANGUAGE plpgsql;</li>
<li>\q</li>
</ul>
</blockquote>
</li>
</ul>
</blockquote>
</li>
<li><p class="first">If you do not want to insert your password everytime to connect to a PostgreSQL database, use these commands (recommended):</p>
<blockquote>
<ul>
<li><p class="first">touch ~/.pgpass</p>
</li>
<li><p class="first">chmod 600 ~/.pgpass</p>
</li>
<li><p class="first">cat &gt;&gt; ~/.pgpass &lt;&lt;EOF</p>
<blockquote>
<ul class="simple">
<li>localhost:*:ppi:ppi:ppi</li>
<li>EOF</li>
</ul>
</blockquote>
</li>
</ul>
</blockquote>
</li>
<li><p class="first">Create the tables which are needed for uploading the results. Change into scripts/ppi-benchmark/Database and execute the following command:</p>
<blockquote>
<ul class="simple">
<li>psql -h localhost -d ppi -U ppi -f init-ppiCV.sql</li>
</ul>
</blockquote>
</li>
</ul>
</div>
</div>
<div class="section" id="how-to-run-the-kernel-pipelines">
<h1>How to run the Kernel Pipelines</h1>
<ul>
<li><p class="first">This section describes how to use run the APG and SL kernel with the combined data set (CPI-DS), CPI-DS_IV, and CPI-DS_NIV in different modes:</p>
<blockquote>
<ul class="simple">
<li>CV: 10x-cross-validation</li>
<li>PR: prediction, based on the trained model of the combined data set (CPI-DS)</li>
<li>XX: cross-corpus training and prediction on user-specific data sets</li>
</ul>
</blockquote>
</li>
<li><p class="first">Furthermore, it contains a short paragraph about how to use these models.</p>
</li>
</ul>
<div class="section" id="apg-kernel-pipeline">
<h2>APG Kernel pipeline</h2>
<ul>
<li><p class="first">These are the 3 main commands:</p>
<blockquote>
<ul class="simple">
<li>CV: make experiment Kernel=APG expTyp=CV InputFile=CPI-DS.xml Processors=4</li>
<li>PR: make experiment Kernel=APG expTyp=PR InputFile=CPI-DS.xml Processors=4</li>
<li>XX: make experiment Kernel=APG expTyp=XX TrainFile=train.xml TestFile=test.xml Processors=4</li>
</ul>
</blockquote>
</li>
<li><p class="first">You can use the test data set DS-40.xml with 40 sentences to check whether your pipeline works.</p>
</li>
<li><p class="first">While the complete CPI-DS_IV pipeline runs around 3:15 h, the test case takes only a few minutes on a notebook with an Intel Core i5-3570 (4x 3.40GHz).</p>
</li>
<li><p class="first">To use this test data set, go to your (new) working directory into &quot;scripts&quot; and run one of the given make commands with a data set from the folder &quot;scripts/generate_XML_files/DS&quot;, e.g.:</p>
<blockquote>
<ul class="simple">
<li>make experiment Kernel=APG expTyp=PR InputFile=annotate_res.xml Processors=4</li>
</ul>
</blockquote>
</li>
<li><p class="first">The make command also uploads the results to the PostgreSQL database.</p>
</li>
<li><p class="first">You can clean your APG workspace after a calculation is finished:</p>
<blockquote>
<ul class="simple">
<li>make clean-APG</li>
</ul>
</blockquote>
</li>
<li><p class="first">Or you can clean the whole workspace:</p>
<blockquote>
<ul class="simple">
<li>make clean-all</li>
</ul>
</blockquote>
</li>
<li><p class="first">Elham Abbasian was involved in creating the shell script for this pipeline as part of her Master Thesis, supervised by Kersten Döring.</p>
</li>
<li><p class="first">Ammar Qaseem updated and completely refined the first version of this pipeline to be used in three modes (cross-validation, prediction, cross-corpus) with only one script APG_Pipeline.sh.</p>
</li>
<li><p class="first">The threshold for a positive or negative prediction is optimized for the highest F1 score and stored in the SQL database.</p>
</li>
<li><p class="first">In &quot;scripts/ppi-benchmark/Experiments/APG/PR/predict/CPI-DS/train0000/predict1.out&quot;, you will find the test predictions with the original class in the second column and the complete sentence identifier in the first column (for the lambda value 1).</p>
</li>
<li><p class="first">In &quot;scripts/ppi-benchmark/Experiments/APG/PR/predict/CPI-DS/train0000/threshold1.out&quot;, you will find the train predictions (self-prediction) for the F1 score optimization with the original class in the second column and the complete sentence identifier in the first column (for the lambda value 1).</p>
</li>
<li><p class="first">The file scripts/ppi-benchmark/Experiments/APG/PR/output.sql contains all prediction values (0 for false, and 1 for true, and the prediction value itself).</p>
</li>
</ul>
</div>
<div class="section" id="sl-kernel-pipeline">
<h2>SL Kernel pipeline</h2>
<ul>
<li><p class="first">These are the 3 main commands:</p>
<blockquote>
<ul class="simple">
<li>CV: make experiment Kernel=SL expTyp=CV InputFile=CPI-DS.xml Processors=4</li>
<li>PR: make experiment Kernel=SL expTyp=PR InputFile=CPI-DS.xml Processors=4</li>
<li>XX: make experiment Kernel=SL expTyp=XX TrainFile=train.xml TestFile=test.xml Processors=4</li>
</ul>
</blockquote>
</li>
<li><p class="first">You can clean your SL workspace after a calculation is finished:</p>
<blockquote>
<ul class="simple">
<li>make clean-SL</li>
</ul>
</blockquote>
</li>
<li><p class="first">Or you can clean the whole workspace:</p>
<blockquote>
<ul class="simple">
<li>make clean-all</li>
</ul>
</blockquote>
</li>
<li><p class="first">The input files need to be located in &quot;scripts/generate_XML_files/DS&quot;.</p>
</li>
<li><p class="first">Ammar Qaseem updated and completely refined the first version of this pipeline to be used in three modes (cross-validation, prediction, cross-corpus) with only one script SL_Pipeline.sh.</p>
</li>
<li><p class="first">Michael Becer was involved in developing a previous version of an independently customized pipeline using the SL kernel as part of his Bachelor Thesis, supervised by Kersten Döring.</p>
</li>
<li><p class="first">Kevin Selm wrote a bugfix for the jSRE software, because it was not possible to use different parameter selections of n and w.</p>
<blockquote>
<ul>
<li><p class="first">Original software link:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="https://hlt-nlp.fbk.eu/technologies/jsre">https://hlt-nlp.fbk.eu/technologies/jsre</a></li>
</ul>
</blockquote>
</li>
</ul>
</blockquote>
</li>
<li><p class="first">The ppi-benchmark pipeline was changed to make use of the JAR file scripts/ppi-benchmark/Kernels/jsre/source/dist/runTrain.jar, which was exported from a newly created Eclipse project with the source code of the original software and the debugged code of Kevin Selm.</p>
<blockquote>
<ul class="simple">
<li>You can find these files in the folder scripts/jSRE_debug.</li>
</ul>
</blockquote>
</li>
<li><p class="first">The complete CPI-DS_IV pipeline runs in around 0:14 h with an Intel Core i5-3570 (4x 3.40GHz).</p>
</li>
<li><p class="first">The threshold for a positive or negative prediction is zero and stored in the SQL database.</p>
</li>
<li><p class="first">In &quot;scripts/ppi-benchmark/Experiments/SL/PR/predict/CPI-DS/train0000/predictn=3w=1.out&quot;, you will find the predictions with the original class in the second column and the complete sentence identifier in the first column (for n=3 and w=1).</p>
</li>
<li><p class="first">The file scripts/ppi-benchmark/Experiments/SL/PR/output.sql contains all prediction values (0 for false, and 1 for true, and the prediction value itself).</p>
</li>
</ul>
</div>
<div class="section" id="data-set-evaluation">
<h2>Data set evaluation</h2>
<ul>
<li><p class="first">The folder results/summary/<em>&lt;nameOfKernel&gt;</em>/<em>&lt;nameOfDataset&gt;</em>/output/ contains scripts to evaluate the results (numbers shown in the tables in the section &quot;Introduction&quot;).</p>
</li>
<li><p class="first">If you want to repeat the evaluation procedure, execute the following steps:</p>
<blockquote>
<ul class="simple">
<li>Remove all files in the folder output, except average.py and header.py and then run:</li>
<li>./get_results.sh</li>
</ul>
</blockquote>
</li>
<li><p class="first"><em>nameOfKernel</em> is the name of the kernel (SL or APG) and <em>nameOfDataset</em> is the name of the dataset(CPI-DS, CPI-DS_IV, or CPI-DS_NIV) .</p>
</li>
<li><p class="first">Check the files CPI-DS_average_header.csv, CPI-DS_IV_average_header.csv, and CPI-DS_NIV_average_header.csv. They exist as a backup in the folder results/summary/<em>&lt;nameOfKernel&gt;</em>/<em>&lt;nameOfDataset&gt;</em>/final/ and the selected SQL results are stored as a backup in the folder results/summary/<em>&lt;nameOfKernel&gt;</em>/<em>&lt;nameOfDataset&gt;</em>/backup_original.</p>
</li>
<li><p class="first">You can reproduce any APG cross-validation run by commenting out lines 188 and 189 in scripts/APG_Pipeline.sh and copying your selected cross-validation splits to CPI-corpora-preparing/splitting/DS.</p>
</li>
<li><p class="first">You can reproduce any SL cross-validation run by commenting out lines 205-207 in scripts/ppi-benchmark/Corpora/Makefile and line 93 in scripts/SL_Pipeline.sh. You also need to copy your selected cross-validation splits to ppi-benchmark/Corpora/Splits/DS.</p>
</li>
<li><p class="first">If you want to reproduce the cross-validation results of CPI-DS_IV and CPI-DS_NIV, please, have a look at the readme files in &quot;scripts/splitting/scripts_for_use_same_splits&quot;.</p>
</li>
</ul>
</div>
<div class="section" id="usage-of-created-models">
<h2>Usage of Created Models</h2>
<ul>
<li><p class="first">If you want to use the models created with any data set, use the XX mode. If you want to use our representative model of PubMed, run the PR mode.</p>
</li>
<li><p class="first">You can use PubMedPortable with its named entity recognition interfaces to prepare sentences with highlighted compounds and proteins.</p>
</li>
<li><p class="first">The basic input for the script annotatedsen_to_xml.py (training_dataset_sorted.csv) can be created by generating a tab-separated format which contains the following columns:</p>
<blockquote>
<ul>
<li><p class="first">PubMed ID</p>
</li>
<li><p class="first">Sentence with XML tags for all named entities</p>
</li>
<li><p class="first">As many more columns as there are pairs of related entities</p>
<blockquote>
<ul class="simple">
<li>Format of each pair: &lt;entity&gt;__&lt;entity&gt;__&lt;interaction&gt;</li>
</ul>
</blockquote>
</li>
</ul>
</blockquote>
</li>
<li><p class="first">This format will automatically be generated with the following example command in the folder &quot;scripts/annotate_entities&quot; (further explanations in the &quot;scripts/annotate_entities/how_to&quot; file):</p>
<blockquote>
<ul class="simple">
<li>make annotate InputFile=pmid_example/pmid_example OutputFile=annotate_res.txt Processors=2</li>
</ul>
</blockquote>
</li>
<li><p class="first">You will need to configure python-nltk to download the punkt tokenizer before:</p>
<blockquote>
<ul class="simple">
<li>cd ~</li>
<li>mkdir nltk_data</li>
<li>ipython</li>
<li>import nltk</li>
<li>nltk.download_shell()</li>
<li>d</li>
<li>punkt</li>
<li>q</li>
</ul>
</blockquote>
</li>
<li><p class="first">The example output of this file, converted to an XML file (with the script annotatedsen_to_xml.py as described in the section &quot;Technical background of the XML data set&quot;), is &quot;scripts/generate_XML_files/DS/annotate_res.xml&quot;.</p>
</li>
<li><p class="first">Considering the output of such an experiment, all positively predicted pairs of entities can be used for an ongoing analysis, e.g. in the process of filtering out interaction partners from large-scale corpora.</p>
</li>
<li><p class="first">If you want to run your own model in the PR mode, you need to copy files from your XX run to specific directories.</p>
</li>
<li><p class="first">In the case of the SL kernel, you need to copy the .model file from &quot;scripts/ppi-benchmark/Experiments/SL/XX/trained/DS/train0&quot; to &quot;scripts/training_model/SL_PR_training/trained_model/DS/train0&quot;.</p>
</li>
<li><p class="first">In the case of the APG kernel, there are more steps:</p>
<blockquote>
<ul class="simple">
<li>Copy your .model file from &quot;scripts/ppi-benchmark/Experiments/APG/XX/trained/DS/train0&quot; to &quot;scripts/training_model/APG_PR_training/trained_model/DS/train0&quot;.</li>
<li>Copy &quot;scripts/ppi-benchmark/Experiments/APG/XX/corpus/DS/train0.txt.gz&quot; to &quot;scripts/training_model/APG_PR_training/corpus_train0.txt.gz&quot;.</li>
<li>Copy &quot;scripts/ppi-benchmark/Experiments/APG/XX/dictionary/DS/train0.txt.gz&quot; to &quot;scripts/training_model/APG_PR_training/dict_train0.txt.gz&quot;.</li>
<li>Copy &quot;scripts/ppi-benchmark/Experiments/APG/XX/linearized/DS/train0.txt.gz&quot; to &quot;scripts/training_model/APG_PR_training/linearized_train0.txt.gz&quot;.</li>
<li>Copy &quot;scripts/ppi-benchmark/Experiments/APG/XX/normalized/DS/train0.txt.gz&quot; to &quot;scripts/training_model/APG_PR_training/normalized_train0.txt.gz&quot;.</li>
</ul>
</blockquote>
</li>
<li><p class="first">Running both kernels in PR mode, one after the other, will generate the files &quot;scripts/ppi-benchmark/Experiments/SL/PR/output.sl.xml&quot; and &quot;scripts/ppi-benchmark/Experiments/APG/PR/output.apg.xml&quot;.</p>
</li>
<li><p class="first">Analogously, the same happens in the XX mode.</p>
</li>
<li><p class="first">Regarding the XX mode, the default parameter for the APG kernel is c=1. You can change it by entering your desired value to &quot;scripts/Run_APG_Kernel.py&quot; in line 257, where c0 = 0 will be used as 2^0 = 1. If you want your model to take e.g. c=2, you have to set c0 = 1, because 2^1 = 2.</p>
</li>
<li><p class="first">You can do the same for the XX mode with the SL kernel by changing &quot;scripts/Run_SL_Kernel.py&quot; lines 88 and 90. The parameters w0 and n0 will be taken as you set them. The default is n=3 and w=1.</p>
</li>
<li><p class="first">These files will have the same content as the input xml file, e.g. annotate_res.xml, except that each positively predicted compound-protein pair will be annotated with 'interaction=&quot;True&quot;'.</p>
</li>
<li><p class="first">All compound-protein pairs which were predicted as non-functional relationships will keep their default annotation 'interaction=&quot;False&quot;'.</p>
</li>
<li><p class="first">The comparison of the predictions of both kernels (output.sl.xml and output.apg.xml) can be used to make a jury decision, resulting in a high confidence for identical outputs.</p>
</li>
</ul>
</div>
</div>
<div class="section" id="large-scale-dataset-application">
<h1>Large-scale dataset application</h1>
<p>The kernels have been successfully applied to all PubMed titles and abstracts that were published before 2018. The full output (in xml format) can be found here:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="ftp://132.230.56.164/CPI/">ftp://132.230.56.164/CPI/</a></li>
</ul>
</blockquote>
<p>It contains one zip file for the predictions of each of the two kernels (SL and APG) as well as one zip file with the combination of both kernels based on a jury decision, i.e. only those relations that were predicted as a functional relation by both kernels were classified to be positive.</p>
</div>
<div class="section" id="contact">
<h1>Contact</h1>
<ul>
<li><p class="first">Please, write an e-mail, if you have questions, feedback, improvements, or new ideas:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="mailto:kersten.doering&#64;gmail.com">kersten.doering&#64;gmail.com</a></li>
<li><a class="reference external" href="mailto:ammar.qaseem&#64;pharmazie.uni-freiburg.de">ammar.qaseem&#64;pharmazie.uni-freiburg.de</a></li>
</ul>
</blockquote>
</li>
<li><p class="first">If you are interested in related projects, visit our working group's homepage:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="http://www.pharmbioinf.uni-freiburg.de/">http://www.pharmbioinf.uni-freiburg.de/</a></li>
</ul>
</blockquote>
</li>
</ul>
<div class="section" id="license">
<h2>License</h2>
<ul class="simple">
<li>The CPI-Pipeline project is published with an ISC license given in &quot;license.txt&quot;.</li>
</ul>
</div>
</div>
</div>
</body>
</html>
